New shadowing rule, part 1

slavapestov opened the pull request:
Module name lookup has its own set of shadowing rules based on the import graph. If a module A imports a module B, and module B imports module C, then when referenced from A, top-level declarations in B would shadow those in C. This shadowing behaves differently depending on the declaration kind:

1) types and variables shadow all symbols with the same name that appear subsequently in the import graph traversal
2) top-level functions shadow all top level functions with the same exact type only

I would like to replace module name lookup shadowing with the shadowing rules used for qualified lookup since this enables some optimizations to be implemented in module name lookup which are currently difficult to do.

As a first step toward this goal, this PR re-implements the module name lookup shadowing rule inside the normal `removeShadowedDecls()` path. This is done by introducing a new ImportCache utility class, which caches the results of various queries against the module import graph.

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test compiler performance

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test compiler performance

slavapestov commented:
# Summary for master full

**Unexpected test results, excluded stats for RxCocoa, SwifterSwift, Base64CoderSwiftUI**

**Regressions found (see below)**

- [Debug-batch](#debug-batch-6d3a1f3bf8f38bce)
    - [debug-batch brief](#debug-batch-brief-6d3a1f3bf8f38bce)
    - [debug-batch detailed](#debug-batch-detailed-6d3a1f3bf8f38bce)
- [Release](#release-6d3a1f3bf8f38bce)
    - [release brief](#release-brief-6d3a1f3bf8f38bce)
    - [release detailed](#release-detailed-6d3a1f3bf8f38bce)



<a name="debug-batch-6d3a1f3bf8f38bce">

# Debug-batch

<a name="debug-batch-brief-6d3a1f3bf8f38bce">

## debug-batch brief
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (3)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 38,842,512,153,491 | 38,725,269,417,877 | -117,242,735,614 | -0.3%
LLVM.NumLLVMBytesOutput | 1,788,495,456 | 1,788,494,784 | -672 | -0.0%
time.swift-driver.wall | 2980.3s | 2970.9s | -9.3s | -0.31%
</details>

<a name="debug-batch-detailed-6d3a1f3bf8f38bce">

## debug-batch detailed
<details>
<summary>Regressed (1)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Sema.OverriddenDeclsRequest | 2,210,158 | 2,249,427 | 39,269 | 1.78% :no_entry:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (151)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumASTBytesAllocated | 51,798,616,599 | 51,808,011,762 | 9,395,163 | 0.02%
AST.NumDecls | 140,002 | 140,002 | 0 | 0.0%
AST.NumDependencies | 207,421 | 207,425 | 4 | 0.0%
AST.NumInfixOperators | 53,782 | 53,782 | 0 | 0.0%
AST.NumLinkLibraries | 0 | 0 | 0 | 0.0%
AST.NumLoadedModules | 270,244 | 270,244 | 0 | 0.0%
AST.NumLocalTypeDecls | 253 | 253 | 0 | 0.0%
AST.NumLookupInModule | 5,437,191 | 5,430,690 | -6,501 | -0.12%
AST.NumLookupQualifiedInAnyObject | 281 | 281 | 0 | 0.0%
AST.NumLookupQualifiedInModule | 2,655,379 | 2,649,301 | -6,078 | -0.23%
AST.NumLookupQualifiedInNominal | 7,320,538 | 7,321,348 | 810 | 0.01%
AST.NumModuleLookupClassMember | 15,754 | 15,754 | 0 | 0.0%
AST.NumModuleLookupValue | 38,532,247 | 38,524,823 | -7,424 | -0.02%
AST.NumObjCMethods | 24,701 | 24,701 | 0 | 0.0%
AST.NumPostfixOperators | 23 | 23 | 0 | 0.0%
AST.NumPrecedenceGroups | 25,668 | 25,668 | 0 | 0.0%
AST.NumPrefixOperators | 99 | 99 | 0 | 0.0%
AST.NumReferencedDynamicNames | 189 | 189 | 0 | 0.0%
AST.NumReferencedMemberNames | 6,322,382 | 6,322,382 | 0 | 0.0%
AST.NumReferencedTopLevelNames | 477,003 | 477,003 | 0 | 0.0%
AST.NumSourceBuffers | 302,844 | 302,844 | 0 | 0.0%
AST.NumSourceLines | 4,685,544 | 4,685,544 | 0 | 0.0%
AST.NumSourceLinesPerSecond | 4,714,873 | 4,689,832 | -25,041 | -0.53%
AST.NumTotalClangImportedEntities | 4,987,430 | 4,988,106 | 676 | 0.01%
AST.NumUnqualifiedLookup | 3,605,519 | 3,605,087 | -432 | -0.01%
Driver.ChildrenMaxRSS | 227,006,955,520 | 227,983,867,904 | 976,912,384 | 0.43%
Driver.DriverDepCascadingDynamic | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingExternal | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingMember | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingNominal | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingTopLevel | 0 | 0 | 0 | 0.0%
Driver.DriverDepDynamic | 0 | 0 | 0 | 0.0%
Driver.DriverDepExternal | 0 | 0 | 0 | 0.0%
Driver.DriverDepMember | 0 | 0 | 0 | 0.0%
Driver.DriverDepNominal | 0 | 0 | 0 | 0.0%
Driver.DriverDepTopLevel | 0 | 0 | 0 | 0.0%
Driver.NumDriverJobsRun | 27,904 | 27,904 | 0 | 0.0%
Driver.NumDriverJobsSkipped | 0 | 0 | 0 | 0.0%
Driver.NumDriverPipePolls | 78,662 | 78,404 | -258 | -0.33%
Driver.NumDriverPipeReads | 69,402 | 69,098 | -304 | -0.44%
Driver.NumProcessFailures | 0 | 0 | 0 | 0.0%
Frontend.MaxMallocUsage | 952,802,772,272 | 952,497,867,648 | -304,904,624 | -0.03%
Frontend.NumInstructionsExecuted | 38,842,512,153,491 | 38,725,269,417,877 | -117,242,735,614 | -0.3%
Frontend.NumProcessFailures | 0 | 0 | 0 | 0.0%
IRModule.NumIRAliases | 197,197 | 197,197 | 0 | 0.0%
IRModule.NumIRBasicBlocks | 6,800,554 | 6,800,554 | 0 | 0.0%
IRModule.NumIRComdatSymbols | 0 | 0 | 0 | 0.0%
IRModule.NumIRFunctions | 3,364,031 | 3,364,031 | 0 | 0.0%
IRModule.NumIRGlobals | 3,522,318 | 3,522,318 | 0 | 0.0%
IRModule.NumIRIFuncs | 0 | 0 | 0 | 0.0%
IRModule.NumIRInsts | 86,650,634 | 86,650,634 | 0 | 0.0%
IRModule.NumIRNamedMetaData | 134,220 | 134,220 | 0 | 0.0%
IRModule.NumIRValueSymbols | 6,231,344 | 6,231,344 | 0 | 0.0%
LLVM.NumLLVMBytesOutput | 1,788,495,456 | 1,788,494,784 | -672 | -0.0%
Parse.NumFunctionsParsed | 267,971 | 267,971 | 0 | 0.0%
Parse.NumIterableDeclContextParsed | 588,486 | 588,503 | 17 | 0.0%
SILModule.NumSILGenDefaultWitnessTables | 0 | 0 | 0 | 0.0%
SILModule.NumSILGenFunctions | 1,709,475 | 1,709,475 | 0 | 0.0%
SILModule.NumSILGenGlobalVariables | 53,884 | 53,884 | 0 | 0.0%
SILModule.NumSILGenVtables | 18,576 | 18,576 | 0 | 0.0%
SILModule.NumSILGenWitnessTables | 72,169 | 72,169 | 0 | 0.0%
SILModule.NumSILOptDefaultWitnessTables | 0 | 0 | 0 | 0.0%
SILModule.NumSILOptFunctions | 2,441,630 | 2,441,630 | 0 | 0.0%
SILModule.NumSILOptGlobalVariables | 55,508 | 55,508 | 0 | 0.0%
SILModule.NumSILOptVtables | 31,332 | 31,332 | 0 | 0.0%
SILModule.NumSILOptWitnessTables | 157,752 | 157,752 | 0 | 0.0%
Sema.AccessLevelRequest | 10,776,479 | 10,774,684 | -1,795 | -0.02%
Sema.AttachedFunctionBuilderRequest | 3 | 3 | 0 | 0.0%
Sema.AttachedPropertyWrapperTypeRequest | 515,019 | 515,019 | 0 | 0.0%
Sema.AttachedPropertyWrappersRequest | 2,098,437 | 2,098,437 | 0 | 0.0%
Sema.CollectOverriddenDeclsRequest | 6,399,987 | 6,394,341 | -5,646 | -0.09%
Sema.CursorInfoRequest | 0 | 0 | 0 | 0.0%
Sema.CustomAttrNominalRequest | 3 | 3 | 0 | 0.0%
Sema.DefaultAndMaxAccessLevelRequest | 52,752 | 52,744 | -8 | -0.02%
Sema.DefaultDefinitionTypeRequest | 7,678 | 7,678 | 0 | 0.0%
Sema.DefaultTypeRequest | 458,934 | 458,934 | 0 | 0.0%
Sema.EmittedMembersRequest | 25,756 | 25,756 | 0 | 0.0%
Sema.EnumRawTypeRequest | 21,253 | 21,253 | 0 | 0.0%
Sema.ExistentialConformsToSelfRequest | 18,532 | 18,499 | -33 | -0.18%
Sema.ExistentialTypeSupportedRequest | 15,729 | 15,729 | 0 | 0.0%
Sema.ExtendedNominalRequest | 5,132,700 | 5,134,359 | 1,659 | 0.03%
Sema.FunctionBuilderTypeRequest | 3 | 3 | 0 | 0.0%
Sema.GetDestructorRequest | 26,296 | 26,296 | 0 | 0.0%
Sema.HasDynamicMemberLookupAttributeRequest | 0 | 0 | 0 | 0.0%
Sema.InheritedDeclsReferencedRequest | 4,998,797 | 5,001,100 | 2,303 | 0.05%
Sema.InheritedTypeRequest | 288,571 | 288,412 | -159 | -0.06%
Sema.InitKindRequest | 102,150 | 102,150 | 0 | 0.0%
Sema.IsAccessorTransparentRequest | 308,581 | 308,581 | 0 | 0.0%
Sema.IsDeclApplicableRequest | 0 | 0 | 0 | 0.0%
Sema.IsDynamicRequest | 1,591,894 | 1,591,894 | 0 | 0.0%
Sema.IsFinalRequest | 2,395,459 | 2,393,798 | -1,661 | -0.07%
Sema.IsGetterMutatingRequest | 422,917 | 422,917 | 0 | 0.0%
Sema.IsImplicitlyUnwrappedOptionalRequest | 2,361,405 | 2,360,647 | -758 | -0.03%
Sema.IsObjCRequest | 1,507,926 | 1,507,883 | -43 | -0.0%
Sema.IsSetterMutatingRequest | 343,444 | 343,444 | 0 | 0.0%
Sema.LazyStoragePropertyRequest | 2,155 | 2,155 | 0 | 0.0%
Sema.MangleLocalTypeDeclRequest | 506 | 506 | 0 | 0.0%
Sema.NamedLazyMemberLoadFailureCount | 17,709 | 17,724 | 15 | 0.08%
Sema.NamedLazyMemberLoadSuccessCount | 27,362,062 | 27,362,661 | 599 | 0.0%
Sema.NominalTypeLookupDirectCount | 34,725,253 | 34,724,622 | -631 | -0.0%
Sema.NumAccessorBodiesSynthesized | 189,241 | 189,241 | 0 | 0.0%
Sema.NumAccessorsSynthesized | 277,024 | 277,024 | 0 | 0.0%
Sema.NumConformancesDeserialized | 8,029,650 | 8,030,101 | 451 | 0.01%
Sema.NumConstraintScopes | 26,975,750 | 26,974,791 | -959 | -0.0%
Sema.NumConstraintsConsideredForEdgeContraction | 86,008,671 | 86,008,551 | -120 | -0.0%
Sema.NumCyclicOneWayComponentsCollapsed | 0 | 0 | 0 | 0.0%
Sema.NumDeclsDeserialized | 61,709,893 | 61,710,029 | 136 | 0.0%
Sema.NumDeclsTypechecked | 1,403,742 | 1,403,742 | 0 | 0.0%
Sema.NumDeclsValidated | 2,411,843 | 2,411,850 | 7 | 0.0%
Sema.NumFunctionsTypechecked | 527,252 | 527,252 | 0 | 0.0%
Sema.NumGenericSignatureBuilders | 1,209,667 | 1,209,765 | 98 | 0.01%
Sema.NumLazyGenericEnvironments | 11,411,890 | 11,416,973 | 5,083 | 0.04%
Sema.NumLazyGenericEnvironmentsLoaded | 300,063 | 299,919 | -144 | -0.05%
Sema.NumLazyIterableDeclContexts | 7,424,290 | 7,426,818 | 2,528 | 0.03%
Sema.NumLazyRequirementSignatures | 740,038 | 740,127 | 89 | 0.01%
Sema.NumLazyRequirementSignaturesLoaded | 499,782 | 499,889 | 107 | 0.02%
Sema.NumLeafScopes | 17,344,928 | 17,344,253 | -675 | -0.0%
Sema.NumTypesDeserialized | 18,385,505 | 18,385,233 | -272 | -0.0%
Sema.NumTypesValidated | 1,620,148 | 1,620,127 | -21 | -0.0%
Sema.NumUnloadedLazyIterableDeclContexts | 4,507,255 | 4,509,011 | 1,756 | 0.04%
Sema.OpaqueReadOwnershipRequest | 262,198 | 262,198 | 0 | 0.0%
Sema.PropertyWrapperBackingPropertyInfoRequest | 510,024 | 510,024 | 0 | 0.0%
Sema.PropertyWrapperBackingPropertyTypeRequest | 515,019 | 515,019 | 0 | 0.0%
Sema.PropertyWrapperMutabilityRequest | 600,401 | 600,401 | 0 | 0.0%
Sema.PropertyWrapperTypeInfoRequest | 1 | 1 | 0 | 0.0%
Sema.ProtocolRequiresClassRequest | 70,227 | 70,089 | -138 | -0.2%
Sema.ProvideDefaultImplForRequest | 6,399,987 | 6,394,341 | -5,646 | -0.09%
Sema.RangeInfoRequest | 0 | 0 | 0 | 0.0%
Sema.RequirementRequest | 105,708 | 105,701 | -7 | -0.01%
Sema.RequirementSignatureRequest | 575,061 | 575,049 | -12 | -0.0%
Sema.RequiresOpaqueAccessorsRequest | 1,295,543 | 1,295,543 | 0 | 0.0%
Sema.RequiresOpaqueModifyCoroutineRequest | 252,660 | 252,660 | 0 | 0.0%
Sema.ResilienceExpansionRequest | 1,671,280 | 1,671,263 | -17 | -0.0%
Sema.ResolveProtocolNameRequest | 0 | 0 | 0 | 0.0%
Sema.RootAndResultTypeOfKeypathDynamicMemberRequest | 0 | 0 | 0 | 0.0%
Sema.RootTypeOfKeypathDynamicMemberRequest | 0 | 0 | 0 | 0.0%
Sema.SelfAccessKindRequest | 4,883,446 | 4,888,094 | 4,648 | 0.1%
Sema.SelfBoundsFromWhereClauseRequest | 6,644,312 | 6,646,577 | 2,265 | 0.03%
Sema.SetterAccessLevelRequest | 147,336 | 147,336 | 0 | 0.0%
Sema.StorageImplInfoRequest | 1,158,037 | 1,158,037 | 0 | 0.0%
Sema.StoredPropertiesAndMissingMembersRequest | 29,852 | 29,852 | 0 | 0.0%
Sema.StoredPropertiesRequest | 287,532 | 287,532 | 0 | 0.0%
Sema.StructuralTypeRequest | 0 | 0 | 0 | 0.0%
Sema.SuperclassDeclRequest | 391,144 | 391,033 | -111 | -0.03%
Sema.SuperclassTypeRequest | 49,019 | 49,019 | 0 | 0.0%
Sema.SynthesizeAccessorRequest | 277,024 | 277,024 | 0 | 0.0%
Sema.TypeCheckFunctionBodyUntilRequest | 527,252 | 527,252 | 0 | 0.0%
Sema.TypeDeclsFromWhereClauseRequest | 28,071 | 28,063 | -8 | -0.03%
Sema.TypeRelationCheckRequest | 0 | 0 | 0 | 0.0%
Sema.USRGenerationRequest | 7,487,809 | 7,483,684 | -4,125 | -0.06%
Sema.UnderlyingTypeDeclsReferencedRequest | 243,508 | 243,318 | -190 | -0.08%
</details>

<a name="release-6d3a1f3bf8f38bce">

# Release

<a name="release-brief-6d3a1f3bf8f38bce">

## release brief
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (3)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 47,055,369,000,908 | 47,027,916,519,654 | -27,452,481,254 | -0.06%
LLVM.NumLLVMBytesOutput | 1,541,593,646 | 1,541,601,170 | 7,524 | 0.0%
time.swift-driver.wall | 5216.7s | 5218.5s | 1.8s | 0.04%
</details>

<a name="release-detailed-6d3a1f3bf8f38bce">

## release detailed
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (21)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumLoadedModules | 30,333 | 30,333 | 0 | 0.0%
AST.NumTotalClangImportedEntities | 1,231,537 | 1,231,543 | 6 | 0.0%
IRModule.NumIRBasicBlocks | 5,632,945 | 5,632,945 | 0 | 0.0%
IRModule.NumIRFunctions | 2,856,736 | 2,856,736 | 0 | 0.0%
IRModule.NumIRGlobals | 3,026,851 | 3,026,851 | 0 | 0.0%
IRModule.NumIRInsts | 53,112,848 | 53,112,848 | 0 | 0.0%
IRModule.NumIRValueSymbols | 5,526,332 | 5,526,332 | 0 | 0.0%
LLVM.NumLLVMBytesOutput | 1,541,593,646 | 1,541,601,170 | 7,524 | 0.0%
SILModule.NumSILGenFunctions | 1,191,466 | 1,191,466 | 0 | 0.0%
SILModule.NumSILOptFunctions | 1,681,840 | 1,681,840 | 0 | 0.0%
Sema.NumConformancesDeserialized | 3,930,087 | 3,930,087 | 0 | 0.0%
Sema.NumConstraintScopes | 26,603,733 | 26,603,733 | 0 | 0.0%
Sema.NumDeclsDeserialized | 10,453,700 | 10,453,700 | 0 | 0.0%
Sema.NumDeclsValidated | 1,826,146 | 1,826,146 | 0 | 0.0%
Sema.NumFunctionsTypechecked | 530,874 | 530,874 | 0 | 0.0%
Sema.NumGenericSignatureBuilders | 295,642 | 295,581 | -61 | -0.02%
Sema.NumLazyGenericEnvironments | 2,062,676 | 2,062,676 | 0 | 0.0%
Sema.NumLazyGenericEnvironmentsLoaded | 35,748 | 35,748 | 0 | 0.0%
Sema.NumLazyIterableDeclContexts | 1,330,132 | 1,330,132 | 0 | 0.0%
Sema.NumTypesDeserialized | 5,621,264 | 5,621,264 | 0 | 0.0%
Sema.NumTypesValidated | 947,067 | 947,067 | 0 | 0.0%
</details>

jrose-apple commented on lib/AST/ASTContext.cpp:
```diff
@@ -871,7 +871,8 @@ ClassDecl *ASTContext::get##NAME##Decl() const { \
       /* Note: use unqualified lookup so we find NSError regardless of */ \
       /* whether it's defined in the Foundation module or the Clang */ \
       /* Foundation module it imports. */ \
```
This comment is wrong to begin with; a qualified lookup will look into re-exported modules as well.

slavapestov replied:
Yeah, and this could probably be cleaned up to just call lookupInModule().

jrose-apple replied:
No, that's the one that *doesn't* look in re-exported modules.

slavapestov replied:
Can you double check so I'm not missing something? I thought lookupValue() is the one that only looks in the module itself whereas lookupInModule() does all the extra work of walking the graph...?

jrose-apple commented:
I mean…if we never hit any of the edge cases for this except that suddenly some things start working, maybe we're okay!

jrose-apple commented:
No obvious comments from scanning through the diff.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov committed:
AST: removeShadowedDecls() takes a DeclContext instead of a ModuleDecl

This allows shadowing rules to depend on the current SourceFile and
not just the current ModuleDecl. For now, this isn't actually
taken advantage of.

slavapestov committed:
AST: Introduce a new ImportCache singleton

slavapestov committed:
AST: Use ImportCache to simplify SourceFile::isImportedImplementationOnly()

slavapestov committed:
AST: Use ImportCache to simplify DeclContext::lookupQualified()

slavapestov committed:
AST: Use ImportCache to simplify UnqualifiedLookupFactory::lookForAModuleWithTheGivenName()

slavapestov committed:
AST: Add module visibility shadowing rule to removeShadowedDecls()

This simulates the shadowing done by ModuleNameLookup, which is about to be
removed.

The basic idea is that if a module A imports a module B, and B imports C,
then from A's point of view, top-level declarations from B will shadow
top-level declarations from C.

slavapestov committed:
AST: Don't compute overload signature unless needed in removeShadowedDecls()

jrose-apple reviewed (commented):
All this looks good if it works, comments are only low-level implementation cleanups.

slavapestov committed:
AST: Qualified lookup doesn't need to prune duplicate results

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
Source compat failure is not related.
